core/src/main/java/org/elasticsearch/gateway/ReplicaShardAllocator.java

```diff
@@ -51,6 +50,59 @@ public ReplicaShardAllocator(Settings settings) {
+                ShardRouting shard = it.next();
+                if (shard.primary() == true) {
+                    continue;
+                }
+                if (shard.initializing() == false) {
+                    continue;
+                }
+                if (shard.relocatingNodeId() != null) {
+                    continue;
+                }
+
+                AsyncShardFetch.FetchResult<TransportNodesListShardStoreMetaData.NodeStoreFilesMetaData> shardStores = fetchData(shard, allocation);
+                if (shardStores.hasData() == false) {
+                    logger.trace("{}: fetching new stores for initializing shard", shard);
+                    continue; // still fetching
+                }
+
+                ShardRouting primaryShard = allocation.routingNodes().activePrimary(shard);
+                assert primaryShard != null : "the replica shard can be allocated on at least one node, so there must be an active primary";
+                TransportNodesListShardStoreMetaData.StoreFilesMetaData primaryStore = findStore(primaryShard, allocation, shardStores);
+                if (primaryStore == null || primaryStore.allocated() == false) {
+                    // if we can't find the primary data, it is probably because the primary shard is corrupted (and listing failed)
+                    // we want to let the replica be allocated in order to expose the actual problem with the primary that the replica
+                    // will try and recover from
+                    // Note, this is the existing behavior, as exposed in running CorruptFileTest#testNoPrimaryData
+                    logger.trace("{}: no primary shard store found or allocated, letting actual allocation figure it out", shard);
+                    continue;
+                }
+
+                MatchingNodes matchingNodes = findMatchingNodes(shard, allocation, primaryStore, shardStores);
+                if (matchingNodes.getNodeWithHighestMatch() != null) {
+                    DiscoveryNode currentNode = allocation.nodes().get(shard.currentNodeId());
+                    DiscoveryNode nodeWithHighestMatch = matchingNodes.getNodeWithHighestMatch();
+                    if (currentNode.equals(nodeWithHighestMatch) == false
+                            && matchingNodes.isNodeFullySynced(currentNode) == false
+                            && matchingNodes.isNodeFullySynced(nodeWithHighestMatch) == true) {
+                        // we found a better match that has a full sync id match, the existing allocation is not fully synced
+                        // so we found a better one, cancel this one
+                        it.moveToUnassigned(new UnassignedInfo(UnassignedInfo.Reason.REALOCATED_REPLICA,
+                                "existing allocation of replica to [" + currentNode + "] cancelled, full sync match found on node [" + nodeWithHighestMatch + "]"));
```

Comment:
can we say "sync id match" instead of "full sync". The latter implies also all bytes matched. I think it will be easier to understand.